src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> prev = tail;
+        Node<T> newNode = new Node<T>(prev, value, null);
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        isIndexExist(index);
+        Node<T> newNode = new Node(tail, value, null);
+        Node<T> oldNode = findNodeByIndex(index);
+        if (oldNode == head) {
+            head.prev = newNode;
+            newNode.next = head;
+            head = newNode;
+            size++;
+        } else {
+            Node oldPrevious = oldNode.prev;
+            oldPrevious.next = newNode;
+            oldNode.prev = newNode;
+            newNode.prev = oldPrevious;
+            newNode.next = oldNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexExist(index);
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexExist(index);
+        Node<T> node = findNodeByIndex(index);
```

Comment:
This is code repeating. You always check the index before getting node by index. What you should do is move the check logic INSIDE the get method.

![image](https://github.com/mate-academy/jv-linked-list/assets/20326858/d5a00457-e455-4320-a1ac-064293e99461)